### ISSUE TITLE
Kernel with ti.random() runtime error while using autodiff

### ISSUE THREAD

**Issue by ljcc0930** [taichi-dev/taichi#2304]
```py
import taichi as ti

loss = ti.field(dtype = ti.f32, shape = (), needs_grad = True)
x = ti.field(dtype = ti.f32, shape = (), needs_grad = True)

@ti.kernel
def work():
    loss[None] = x[None] * ti.random()
    # loss[None] = x[None] * 0.5 # this works

x[None] = 10
with ti.Tape(loss):
    work()
print(x.grad[None])
```

**Comment by victoriacity** [comment]
Taichi uses source code transformation ([Life of a Taichi Kernel, Page 57](https://yuanming.taichi.graphics/publication/2020-life-of-kernel/life_of_a_taichi_kernel.pdf)) to obtain gradient kernels. This implies that when differentiating a kernel, all intermediate results for the forward computation need to be recalculated when getting the gradients. 

Say we have a function (written in static single assignment):
```python
def forward(x):
    x0 = x
    x1 = g(x0)
    x2 = h(x1)
    return x2
```
With automatic differentiation the adjoint function is
```python
def adjoint(x, dz):
    x0 = x
    x1 = g(x0)
    x2 = h(x1)
    x2_grad = dz 
    x1_grad = x2_grad * dh(x1)
    x0_grad = x1_grad * dg(x0)
    return x0_grad
```
where dz is the derivative of the loss w.r.t the output of `forward(x)` and the loss w.r.t `x` is subsequently calculated by the chain rule. 

Now if we make `g(x) := x + ti.random()` and `h(x) := sin(x)`, this implies that when calling `forward` and `adjoint` for a given kernel, `ti.random()` should give the same random number. In other words, it requires caching all random numbers generated for differentiable kernel calls. I'm not sure if this is worth implementing, but a more robust way is to pre-generate all random numbers in a separate field.

**Comment by ljcc0930** [comment]
Indeed. Still, a more convenient way is that we can raise a compile error when using ti.random() in Autodiff.
